### ISSUE TITLE
random pull failures when using singularityhub image in travis tests

### ISSUE THREAD

**Issue by ltalirz** [singularityhub/singularityhub.github.io#187]
This is not an issue about building a singularity image, but rather a question concerning pulling an image as part of integration tests on platforms like travis CI.

Just as docker is routinely used in travis tests, I thought I would use a (small) singularity image as part of a test on travis CI.
However, when doing this I run into random failures when pulling the singularity image.

Here is the failed pull
https://travis-ci.org/ltalirz/aiida-zeopp/jobs/560391554#L625
with error message
`Error getting image manifest using url https://www.singularity-hub.org/api/container/ltalirz/singularity-recipe-zeopp:ansible`

Here is the pull working fine (which is the merge commit of the PR above, i.e. no code changes to `.travis.yml` whatsoever):
https://travis-ci.org/ltalirz/aiida-zeopp/jobs/560538474#L617

Am I running into a "pull rate limit" here?
Please let me know whether the practice of using singularity images in continuous integration tests is generally discouraged at this point and, if yes, whether there is a better solution.

P.S. Here the link to the singularity image used: https://singularity-hub.org/collections/2615

**Comment by ltalirz** [comment]
And here is the relevant part of the travis CI configuration
https://github.com/ltalirz/aiida-zeopp/blob/e5f748ab3942affebc395ac3d14c282fec6d2f3e/.travis.yml#L34-L38

**Comment by vsoch** [comment]
hey @ltalirz ! Generally I'd recommend not using an shub URI for a CI pipeline, unless you are certain it's only run a few times a week.  However, you are only at a count of 10, so you are still well under the limit (and I just pulled the container without issue)!

What version of Singularity is being installed via the package manager? (Could it be a version that is slow to pull or has a bug?). Is this the first failure, have you retried it? It's likely an issue of timing - when the request is made to download a container, the exact moment that the manifest is presented, the signed URL is generated and it expires only a few seconds after generation. If Travis (or Singularity) is for some reason slow to grab the "image" field from the manifest and then request it, it would expire and then deny.

Otherwise, your configuration looks okay to me.

**Comment by vsoch** [comment]
Also, the failure you linked is from a month ago. Have you tested since then?

**Comment by ltalirz** [comment]
thanks a lot @vsoch for the superfast reply!

> Generally I'd recommend not using an shub URI for a CI pipeline, unless you are certain it's only run a few times a week. However, you are only at a count of 10, so you are still well under the limit 

Ok. Yes, these tests are not going to be running very frequently.

> (and I just pulled the container without issue)!

I forgot to mention that when I tried pulling from my laptop, I never had any issues.
Is there a limit per IP address?

> What version of Singularity is being installed via the package manager? (Could it be a version that is slow to pull or has a bug?).

It is using `singularity-container amd64 2.6.1-2~nd14.04+1`
I realize there is version 3 out, so you are saying this might be related to the version of the client?
I.e. it is not a server issue?

> Is this the first failure, have you retried it?
> Also, the failure you linked is from a month ago. Have you tested since then?

Yes, I just tried again an hour ago and ran into the issue again.
Note that in the same build there are jobs where it works (the ones that run 3+ minutes), and jobs where it doesn't:
https://travis-ci.org/ltalirz/aiida-zeopp/builds/578967326

> It's likely an issue of timing - when the request is made to download a container, the exact moment that the manifest is presented, the signed URL is generated and it expires only a few seconds after generation. If Travis (or Singularity) is for some reason slow to grab the "image" field from the manifest and then request it, it would expire and then deny.

I see. Is there a way for me to work around that?

**Comment by vsoch** [comment]
I would recommend changing your CI setup entirely - if the repository is meant to test the container, then you would want to *build* it during the run using Singularity, and then on successful merge to master the container will build on Singularity Hub.

If you need to use a container as a base for something, I would instead use a docker uri and pull down to a Singularity container. I can't control the networking of Travis, or the speed of how the older Singularity pulls, so if I were in your shoes, I would do one of the above.

**Comment by ltalirz** [comment]
> If you need to use a container as a base for something, I would instead use a docker uri and pull down to a Singularity container.

Ok, just to make sure I'm getting what you are suggesting:

For every singularity container that I would like to use as part of CI tests, I should create a docker container with a Dockerfile that essentially does the steps that are currently in the travis config
https://github.com/ltalirz/aiida-zeopp/blob/e5f748ab3942affebc395ac3d14c282fec6d2f3e/.travis.yml#L34-L38

Then, on travis I should pull this docker container and run the singularity container inside the docker container.

Is that correct? Or perhaps there is a shortcut here that I am not aware of...

**Comment by ltalirz** [comment]
> I can't control the networking of Travis, or the speed of how the older Singularity pulls

Happy to look into installing version 3 of the client on travis and checking whether the problem goes away.

**Comment by vsoch** [comment]
For your first question, I'm suggesting:

 - you build the container using the Singularity recipe in the CI setup, which is logical because on PR merge it will build the "production" on Singularity Hub after testing. OR
 - yes, provide a Dockerfile that builds the same container, and pull as docker://<container> to singularity for your repository.

I don't mean pulling a Docker container to run Singularity inside of - that doesn't really help you with anything, but adds another layer you have to deal with. 

If you want to try version 3, you can follow the recipe here https://github.com/singularityhub/travis-ci. Another suggestion, since Travis isn't super reliable these days, is to use CircleCI. There is even an Orb that makes it very easy to install Singularity:

 - https://github.com/singularityhub/singularity-orb
 - https://circleci.com/orbs/registry/orb/singularity/singularity

If I've answered your question, feel free to close the issue.

**Comment by ltalirz** [comment]
> yes, provide a Dockerfile that builds the same container, and pull as docker:// to singularity for your repository.

Ok, thanks -  I wasn't aware that singularity can pull from docker hub as well.
Of course, this means I would need to build a docker image for every singularity image that I would like to use in this way.

> If you want to try version 3, you can follow the recipe here https://github.com/singularityhub/travis-ci. 

Thanks. I tried building it locally and it realize it does need quite some dependencies. I guess it is probably not a good idea to build the singularity client from source on every travis job...
I also checked the [apt packages from the neurodebian repo](http://neuro.debian.net/pkgs/singularity-container.html) but for the latest ubuntu version I can get on travis (bionic 18.04), the version is still 2.6

> Another suggestion, since Travis isn't super reliable these days, is to use CircleCI. There is even an Orb that makes it very easy to install Singularity:

Thank you, that is great. Perhaps this is indeed the best way forward.

> If I've answered your question, feel free to close the issue.

Yes, thanks for your help! 

To summarize:
 * I'm likely not hitting a rate limit of the singularity hub server (?)
 * the issue may be related to timing: when the request is made to download a container, the exact moment that the manifest is presented, the signed URL is generated and it expires only a few seconds after generation.
 * upgrading the singularity client from version 2.6 to version 3 may help. as of now, one would need to compile version 3 from source on travis
 * there is an orb for CircleCI that directly provides the client version 3 and would make testing there very straightforward

**Comment by vsoch** [comment]
You could actually use a cache, either with Travis or Circle, so the build would not be so terrible. Yes, let's summarize!

 - You aren't hitting a build limit of the Singularity Hub Server
 - The issue is very likely related to the timing of using the signed URL
 - You could definitely try updating Singularity (if it's Singularity), OR trying a different build service (if it's Travis)
 - yep :)

**Comment by ltalirz** [comment]
> You aren't hitting a build limit of the Singularity Hub Server

Just to add to this - I was wondering whether I could reproduce the issue locally as well, so I wrote this simple bash script:
```bash
singularity pull shub://ltalirz/singularity-recipe-zeopp:ansible --name im1.sif > run_1 2>&1 &
singularity pull shub://ltalirz/singularity-recipe-zeopp:ansible --name im2.sif  > run_2 2>&1 &
singularity pull shub://ltalirz/singularity-recipe-zeopp:ansible --name im3.sif  > run_3 2>&1 &
singularity pull shub://ltalirz/singularity-recipe-zeopp:ansible --name im4.sif  > run_4 2>&1 &
```

On an Ubuntu VM of mine, the first time I ran it, everything worked fine.
The second time I ran it I got 3 "manifest errors" [1] and one success.
The 3rd time I ran it I got 4 "manifest errors".

I don't know but to me this looks like I'm hitting a server limit (I'm running again version 2.6.1 of the client and will try with version 3 as well).

[1]  `Error getting image manifest using url https://www.singularity-hub.org/api/container/ltalirz/singularity-recipe-zeopp:ansible`

**Comment by vsoch** [comment]
Sorry I can't change the url expiration time.

**Comment by vsoch** [comment]
Oh, so for that - what you are hitting is the rate limit of how often you can query - I need to check on what it is, but sequential requests like that would likely trigger it.

**Comment by ltalirz** [comment]
>  what you are hitting is the rate limit of how often you can query - I need to check on what it is, but sequential requests like that would likely trigger it.

Aha! That makes more sense to me.

**Comment by vsoch** [comment]
Yep that's it! You get 10 per day, and then it's a block. The idea is that nobody (should anyway) need to pull a container that many times, especially with cache.

Sorry totally forgot about that limit! Of course.

**Comment by vsoch** [comment]
On the bright side, it's great to know that the limits are working properly! :muscle:

**Comment by ltalirz** [comment]
> You get 10 per day, and then it's a block. The idea is that nobody (should anyway) need to pull a container that many times, especially with cache.

I see. For CI tests that limit is pretty harsh - even when using a cache per build and when running tests only every few days, you're quite likely to run into this limit (depending on the details of how the 10 are computed, even other CI users may be contributing to your rate).

Any chance this limit could be increased a bit for IP ranges of CI services (if such things are known... I would have to check)?

**Comment by vsoch** [comment]
No, I'm really sorry. I had it very generously (not limited) and this was maliciously abused in a way that greatly hurt us. I will consistently maintain this reasonable limit for pulling containers to a cache, and I've made very reasonable suggestions for working around it.

**Comment by ltalirz** [comment]
Ok - so, the conclusion would be that one should not pull images from singularity hub in CI tests (be it on travis, circleci or otherwise), but rather use docker images from docker hub, correct?

No problem - I was just asking for the recommended practice.

**Comment by vsoch** [comment]
Yes, or use the CI to test building the image (which is normally how to go about it). You are doing it a bit backward - having the production image built to singularity hub first, and then testing it with pull. You'd want to build in Travis, ensure that your tests pass, and on merge to master have that deploy the production build. That's the standard workflow for build -> test -> deploy.

**Comment by ltalirz** [comment]
Ok, thanks.

> Yes, or use the CI to test building the image (which is normally how to go about it). You are doing it a bit backward - having the production image built to singularity hub first, and then testing it with pull. 

Just to clarify this final point from my side: I'm *not* testing the image here.
I'm running an integration test of a python code that uses the executable the image provides.

So, I would say pulling the image from somewhere (singularity hub / docker hub / ...) is really preferable to building the image over and over again, since the image does not change.

**Comment by vsoch** [comment]
If you are using the same image, then just add caching to your travis setup. With newer versions of Singularity the image should be cached in $HOME/.singularity.

**Comment by ltalirz** [comment]
Good point. Even with earlier versions, I could download the image to a cached folder and pull only if I don't find the image in that folder. That could work.
